### PR TITLE
[Mono.Data.Sqlite] Added checks for entry point existence

### PR DESCRIPTION
**Pull request by mattleibow**
As a result of https://bugzilla.xamarin.com/show_bug.cgi?id=41723

~~Using `Dlfcn.dlsym` to avoid P/Invoke exceptions, by first checking at startup for the existence of the various entry points. Then, instead of waiting for an exception, check the field to call the correct method the first time.~~

To avoid exceptions, just get the version number once, and then associate the version with API members. No need for entry point lookups, just integer comparison.

Version details: https://www.sqlite.org/c3ref/c_source_id.html
API details per version: https://sqlite.org/changes.html

### CODE REVIEW

**Comment by vargaz** [comment]
ios discourages the use of dlopen()/dlsym () so this might not work in the future.

**Comment by spouliot** [comment]
The API seems to be an addition in 3.5 [1] so it might be possible to call one of the version functions [2] to set the booleans. 

In fact that could become the general check (i.e. not iOS specific) as exceptions are slow and this will also affect older platforms that ships libsqlite (e.g. old OSX and older Linux distros).

[1] https://www.sqlite.org/34to35.html
[2] https://www.sqlite.org/c3ref/libversion.html

**Comment by mattleibow** [comment]
@vargaz @spouliot Thanks for the feedback. I think this will work wonders :) for everyone... ^^^

**Comment by mattleibow** [comment]
I chose to use the `sqlite3_libversion_number` instead of a string as it seemed more reliable. You never know when some custom build adds an extra version or even some non-number character.

File: mcs/class/Mono.Data.Sqlite/Mono.Data.Sqlite_2.0/UnsafeNativeMethods.cs [hunk 1]
```diff
@@ -16,6 +16,29 @@ namespace Mono.Data.Sqlite
...
+      int version = sqlite3_libversion_number();
+      int release = version % 1000;
+      int minor = (version / 1000) % 1000;
+      int major = version / 1000000;
+
+      // set the various versions
+      // https://sqlite.org/changes.html
+      bool v3_5_0 = major >= 3 && minor >= 5 && release >= 0;
+      bool v3_7_14 = major >= 3 && minor >= 7 && release >= 14;
+      bool v3_7_3 = major >= 3 && minor >= 7 && release >= 3;
```
Comment (akoeplinger) [inline mcs/class/Mono.Data.Sqlite/Mono.Data.Sqlite_2.0/UnsafeNativeMethods.cs]: the bools will be false for e.g. version `3.8.0` or `4.0.0`, that looks wrong to me.

Consider using the Version class from the framework:

```
var version = new Version (major, minor, release);
use_sqlite3_open_v2 = version >= new Version (3, 5, 0);
...
```

Comment (mattleibow) [inline mcs/class/Mono.Data.Sqlite/Mono.Data.Sqlite_2.0/UnsafeNativeMethods.cs]: WOW... I missed that logic 😢

**Comment by mattleibow** [comment]
@akoeplinger I am now using the `Version` type to do comparison.

**Comment by spouliot** [comment]
LGTM, we'll need this backported in `mono-4.6.0-branch` for C8. Thanks!

**Comment by mattleibow** [comment]
@spouliot Do I do that, or is that for the team that merges this?

**Comment by akoeplinger** [comment]
@mattleibow I cherry-picked to mono-4.6.0-branch as per @spouliot's comment.

**Comment by spouliot** [comment]
@akoeplinger, @mattleibow thanks to both! :)

**Pull request merged**